src/sql/workbench/contrib/welcome/page/browser/welcomePage.ts

```diff
@@ -0,0 +1,1077 @@
+import { registerColor, focusBorder, descriptionForeground, foreground, contrastBorder, activeContrastBorder } from 'vs/platform/theme/common/colorRegistry';
+import { getExtraColor } from 'vs/workbench/contrib/welcome/walkThrough/common/walkThroughUtils';
+import { IExtensionsWorkbenchService } from 'vs/workbench/contrib/extensions/common/extensions';
+import { IEditorInputFactory, EditorInput } from 'vs/workbench/common/editor';
+import { INotificationService, Severity } from 'vs/platform/notification/common/notification';
+import { TimeoutTimer } from 'vs/base/common/async';
+import { areSameExtensions } from 'vs/platform/extensionManagement/common/extensionManagementUtil';
+import { ILabelService } from 'vs/platform/label/common/label';
+import { IFileService } from 'vs/platform/files/common/files';
+import { ExtensionType } from 'vs/platform/extensions/common/extensions';
+import { IRecentlyOpened, isRecentWorkspace, IRecentWorkspace, IRecentFolder, isRecentFolder, IWorkspacesService } from 'vs/platform/workspaces/common/workspaces';
+import { CancellationToken } from 'vs/base/common/cancellation';
+import { IHostService } from 'vs/workbench/services/host/browser/host';
+import { IProductService } from 'vs/platform/product/common/productService';
+import { KeyCode } from 'vs/base/common/keyCodes';
+import { joinPath } from 'vs/base/common/resources';
+
+
+const configurationKey = 'workbench.startupEditor';
+const oldConfigurationKey = 'workbench.welcome.enabled';
+const telemetryFrom = 'welcomePage';
+
+
+export class WelcomePageContribution implements IWorkbenchContribution {
+	constructor(
+		@IInstantiationService private readonly instantiationService: IInstantiationService,
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		@IEditorService private readonly editorService: IEditorService,
+		@IBackupFileService private readonly backupFileService: IBackupFileService,
+		@IFileService private readonly fileService: IFileService,
+		@IWorkspaceContextService private readonly contextService: IWorkspaceContextService,
+		@ILifecycleService private readonly lifecycleService: ILifecycleService,
+		@ICommandService private readonly commandService: ICommandService,
+	) {
+		this.enableWelcomePage().catch(onUnexpectedError);
+	}
+	private async enableWelcomePage(): Promise<void> {
+		const enabled = isWelcomePageEnabled(this.configurationService, this.contextService);
+		if (enabled && this.lifecycleService.startupKind !== StartupKind.ReloadedWindow) {
+			const hasBackups: boolean = await this.backupFileService.hasBackups();
+			const activeEditor = this.editorService.activeEditor;
+			if (!activeEditor && !hasBackups) {
+				const openWithReadme = this.configurationService.getValue(configurationKey) === 'readme';
+				if (openWithReadme) {
+					let readmes = await Promise.all(this.contextService.getWorkspace().folders.map(async folder => {
```

Comment:
So this is going to be Array<SomeObject | undefined>, just make sure where you're using this later can handle undefined values in the array.

Reply:
We call arrays.coalesceInPlace(readmes) to strip out undefined values afterwards